core/lib/dal/src/storage_logs_dal.rs

```diff
@@ -722,44 +723,62 @@ impl StorageLogsDal<'_, '_> {
-                storage_logs.hashed_key,
-                storage_logs.value,
-                initial_writes.index
-            FROM
-                storage_logs
-            INNER JOIN initial_writes ON storage_logs.hashed_key = initial_writes.hashed_key
-            WHERE
-                storage_logs.miniblock_number <= $1
-                AND storage_logs.hashed_key >= $2::bytea
-                AND storage_logs.hashed_key <= $3::bytea
-            ORDER BY
-                storage_logs.hashed_key
-            "#,
-            i64::from(l2_block_number.0),
-            key_range.start().as_bytes(),
-            key_range.end().as_bytes()
-        )
-        .instrument("get_tree_entries_for_l2_block")
-        .with_arg("l2_block_number", &l2_block_number)
-        .with_arg("key_range", &key_range)
-        .fetch_all(self.storage)
-        .await?;
+        const QUERY_LIMIT: usize = 10_000;
 
-        let rows = rows.into_iter().map(|row| StorageRecoveryLogEntry {
-            key: H256::from_slice(&row.hashed_key),
-            value: H256::from_slice(&row.value),
-            leaf_index: row.index as u64,
-        });
-        Ok(rows.collect())
+        // Break fetching from the DB into smaller chunks to make DB load more uniform.
+        let mut entries = vec![];
+        loop {
+            let rows = sqlx::query!(
+                r#"
+                SELECT
+                    storage_logs.hashed_key,
+                    storage_logs.value,
+                    initial_writes.index
+                FROM
+                    storage_logs
+                INNER JOIN initial_writes ON storage_logs.hashed_key = initial_writes.hashed_key
+                WHERE
+                    storage_logs.miniblock_number <= $1
+                    AND storage_logs.hashed_key >= $2::bytea
+                    AND storage_logs.hashed_key <= $3::bytea
+                ORDER BY
+                    storage_logs.hashed_key
+                LIMIT
+                    $4
+                "#,
+                i64::from(l2_block_number.0),
+                key_range.start().as_bytes(),
+                key_range.end().as_bytes(),
+                QUERY_LIMIT as i32
+            )
+            .instrument("get_tree_entries_for_l2_block")
+            .with_arg("l2_block_number", &l2_block_number)
+            .with_arg("key_range", &key_range)
+            .fetch_all(self.storage)
+            .await?;
+
+            let fetched_count = rows.len();
+            entries.extend(rows.into_iter().map(|row| StorageRecoveryLogEntry {
+                key: H256::from_slice(&row.hashed_key),
+                value: H256::from_slice(&row.value),
+                leaf_index: row.index as u64,
+            }));
+
+            if fetched_count < QUERY_LIMIT {
+                break;
+            }
+            // `unwrap()` is safe: `entries` contains >= QUERY_LIMIT items.
+            let next_key = u256_to_h256(h256_to_u256(entries.last().unwrap().key) + 1);
```

Comment:
nit: it can overflow if `entries.last().unwrap().key` is `U256::MAX`. Cannot happen in practice but I would still add smth like `if entries.last().unwrap().key == U256::MAX { break; }